src/imars3dv2/filters/tilt.py

```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+import numpy as np
+import tomopy.util.mproc as mproc
+import concurrent.futures as cf
+from typing import Tuple
+from functools import partial
+from scipy.optimize import minimize_scalar
+from skimage.transform import rotate
+from skimage.registration import phase_cross_correlation
+
+
+def tilt_correction(
+    arrays: np.ndarray,
+    omegas: np.ndarray,
+    low_bound: float = -5.0,
+    high_bound: float = -5.0,
+    cut_off_angle_deg: float = 1e-3,
+    ncore: int = -1,
+):
+    """
+    Detect and correct the rotation axis tilt from the given radiograph stack.
+
+    Parameters
+    ----------
+    @param arrays:
+        The radiograph stack fro tilt correction
+    @param omegas:
+        The list of omega angles in radians (follow tomopy convention)
+    @param low_bound:
+        The lower bound of the tilt angle search space
+    @param high_bound:
+        The upper bound of the tilt angle search space
+    @param cut_off_angle_deg:
+        The angle in degrees to cut off the rotation axis tilt correction, i.e.
+        skip applying tilt correction for tilt angles that are too small.
+    @param ncore:
+        Number of cores to use for parallel median filtering, default is -1,
+        which means using all available cores.
+
+    Returns
+    -------
+    @return:
+        The tilt corrected array
+    """
+    if arrays.ndim != 3:
+        raise ValueError("arrays must be a 3d array")
+    # step 1: find the 180 degree pairs
+    idx_lowrange, idx_highrange = find_180_deg_pairs_idx(omegas, in_degrees=False)
+    # step 2: calculate the tilt from each pair
+    ncore = mproc.mp.cpu_count() if ncore == -1 else int(ncore)
+    with cf.ProcessPoolExecutor(ncore) as e:
+        jobs = []
+        for low_idx, high_idx in zip(idx_lowrange, idx_highrange):
+            img0 = arrays[low_idx]
+            img1 = arrays[high_idx]
+            jobs.append(e.submit(calculate_tilt, img0, img1, low_bound, high_bound))
+    tilts = np.array([job.result() for job in jobs])
+    # use the average of the found tilt angles
+    tilt = np.mean(tilts)
+    # step 3: apply the tilt correction
+    if abs(tilt) < cut_off_angle_deg:
```

Comment:
If the tilt angle is within range return the input data? This doesn't seem right